feat: Add book service for managing book

Chuckboliver opened the pull request:
- Find all books by author endpoint
- Add book endpoint

Chuckboliver committed:
feat: Add book service for managing book

Chuckboliver committed:
chore: Remove schema.sql in resources

Chuckboliver committed:
chore: Update README.MD documentation

Chuckboliver committed:
fix: Fix typo in README.md

Chuckboliver committed:
fix: Fix integration test property

Chuckboliver committed:
chore: Update README.md

Chuckboliver committed:
chore: Bind failsafe integration-test goal to maven integration-test and verify phases

marcinmilewski93 commented on pom.xml:
```diff
@@ -0,0 +1,100 @@
+    <parent>
+        <groupId>org.springframework.boot</groupId>
+        <artifactId>spring-boot-starter-parent</artifactId>
+        <version>2.7.18</version>
```
This Spring Boot version has some vulnerabilities. Why haven't you decided to use the newer Spring Boot 3?

Chuckboliver replied:
@marcinmilewski93 In my current company, many projects I work on have not yet migrated to Spring Boot 3, so we still use Java 8 and Spring Boot 2. I chose this version because I’m most familiar with it, but I can upgrade to Spring Boot 3 in my free time if needed.

marcinmilewski93 replied:
@Chuckboliver It's okay, I just wanted to understand the reason. However, for your project, I think there would likely be no code changes - just pom.

marcinmilewski93 commented on pom.xml:
```diff
@@ -0,0 +1,100 @@
+    <name>book</name>
+    <description>RESTful API for Managing Books</description>
+    <properties>
+        <java.version>1.8</java.version>
```
Java 8's public support from Oracle ended in 2019, five years after its release. Why haven't you chosen a newer version with broader support?

marcinmilewski93 commented on pom.xml:
```diff
@@ -0,0 +1,100 @@
+        <dependency>
+            <groupId>org.apache.commons</groupId>
+            <artifactId>commons-lang3</artifactId>
+            <version>3.17.0</version>
```
As far as I remember, the Apache Commons Lang3 version is also managed by Spring Boot. Why haven't you let Spring Boot handle it?

marcinmilewski93 commented on pom.xml:
```diff
@@ -0,0 +1,100 @@
+            </plugin>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-failsafe-plugin</artifactId>
```
Maven Failsafe for managing integration tests— I like it👍

marcinmilewski93 commented on schema.sql:
```diff
@@ -0,0 +1,10 @@
+    PRIMARY KEY (id)
+);
+
+CREATE INDEX author_idx on book (author);
```
Every file in this PR that ends with an icon like this 🚫 indicates "No newline at end of file." Could you resolve it? 🙂

marcinmilewski93 commented on src/main/java/com/chuckboliver/bookservice/book/AddBookRequest.java:
```diff
@@ -0,0 +1,19 @@
+package com.chuckboliver.bookservice.book;
```
Packaged by feature! ➕

marcinmilewski93 commented on src/main/java/com/chuckboliver/bookservice/book/AddBookRequest.java:
```diff
@@ -0,0 +1,19 @@
+import javax.validation.constraints.PastOrPresent;
+import java.time.chrono.ThaiBuddhistDate;
+
+@Value
```
Immutable! ➕

marcinmilewski93 commented on src/main/resources/schema.sql:
```diff
@@ -0,0 +1,10 @@
+    PRIMARY KEY (id)
+);
+
+CREATE INDEX author_idx on book (author);
```
Index! ➕

marcinmilewski93 commented on src/main/java/com/chuckboliver/bookservice/common/jackson/ThaiBuddhistDateDeserializer.java:
```diff
@@ -0,0 +1,22 @@
+import java.io.IOException;
+import java.time.chrono.ThaiBuddhistDate;
+
+public class ThaiBuddhistDateDeserializer extends JsonDeserializer<ThaiBuddhistDate> {
```
I like this approach! ➕

marcinmilewski93 commented on src/test/java/com/chuckboliver/bookservice/book/BookControllerIT.java:
```diff
@@ -0,0 +1,185 @@
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
+@ActiveProfiles("it")
+@AutoConfigureMockMvc
+class BookControllerIT {
```
Correct IT suffix, allowing it to be handled by the Failsafe plugin! ➕

marcinmilewski93 commented on pom.xml:
```diff
@@ -0,0 +1,100 @@
+            <artifactId>spring-boot-starter-validation</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>com.h2database</groupId>
```
I was hoping to see Testcontainers, especially with the improved support in Spring Boot 3, but H2 it's still an acceptable choice :)

marcinmilewski93 commented on src/test/java/com/chuckboliver/bookservice/book/BookControllerTest.java:
```diff
@@ -0,0 +1,234 @@
+            verify(bookService).findByAuthor(author);
+        }
+
+        @ParameterizedTest
```
ParameterizedTest! ➕

marcinmilewski93 commented on README.md:
````diff
@@ -1 +1,173 @@
+docker compose up -d
+```
+
+2. Create database schema and initial data with `create-schema.sh`
````
I'm curious why you decided not to use the library like Liquibase or Flyway?

Chuckboliver replied:
@marcinmilewski93 I have never used these libraries before, so I decided not to use them. However, they interest me, so I will learn and try them in my next project 😄.

Chuckboliver committed:
Upgrade project JDK to JDK23 and Spring Boot version 3.4.2

- Ensure every file ends with a line break
- Change @Value class to Java record
- Change deprecated @MockBean to @MockitoBean

Chuckboliver committed:
chore: Update README.md

marcinmilewski93 commented on src/main/java/com/chuckboliver/bookservice/common/ValidationErrorResponse.java:
```diff
@@ -0,0 +1,18 @@
+import java.util.List;
+
+@Value
+public class ValidationErrorResponse {
```
Since you already have a newer version of Java, would you be able to get rid of Lombok from this class and convert it to a record? :)

Chuckboliver replied:
sure

Chuckboliver replied:
@marcinmilewski93 I have converted all the remaining `@Value` class in my recent commit. Please take a look!

marcinmilewski93 replied:
That's exactly what I meant. Great job!

marcinmilewski93 reviewed (approved):
Good job! I like your update from using Lombok's @Value to the new Java Records for the DTO class. I also appreciate how you pay attention to deprecations and resolve IDE warnings. LGTM!

Chuckboliver committed:
fix: Convert remaining @Value class to Java record
